sys/include/net/sock/addr.h

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2016 Freie Universität Berlin
+ *                    Kaspar Schleiser <kaspar@schleiser.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_sock_addr   Address abstractions
+ * @ingroup     net_sock
+ * @brief       Address abstractions for usage with @ref net_sock
+ * @{
+ *
+ * @file
+ * @brief   Address abstraction definitions for @ref net_sock
+ *
+ * @author  Martine Lenders <m.lenders@fu-berlin.de>
+ * @author  Kaspar Schleiser <kaspar@schleiser.de>
+ */
+#ifndef SOCK_ADDR_H_
+#define SOCK_ADDR_H_
+
+#include <stdint.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Special @ref net_netif "netif" ID for "any interface"
+ * @todo    Define in @ref net_netif
+ */
+#define SOCK_ADDR_ANY_NETIF (0)
+
+/**
+ * @brief   Address to bind to any IPv4 address
+ */
+#define SOCK_ADDR_IPV4_ADDR_ANY (0U)
+
+/**
+ * @brief   Address to bind to any IPv6 address
+ */
+#define SOCK_ADDR_IPV6_ADDR_ANY { { 0 } }
+
+typedef uint32_t sock_addr_ipv4_t;
+
+/**
+ * @brief   Type to represent an IPv6 address
+ */
+typedef struct {
+    uint8_t u8[16];         /**< bytes of the address */
+} sock_addr_ipv6_t;
```

Comment:
Anyway, we're running in circles here. I have (strong) dislike of defining redundant data structure, you have a dislike of the current IPv6 header. Possible resolution strategies:
1. I live with yet another IP address data type,
2. you live with including `net/ipv6` here,
3. we define a data type for IPv6 addresses somewhere else, similar to the proposal here and replace both data structures.